src/Access/GSSAcceptor.h

```diff
@@ -0,0 +1,66 @@
+#pragma once
+
+#if !defined(ARCADIA_BUILD)
+#   include "config_core.h"
+#endif
+
+#include <Access/Credentials.h>
+#include <common/types.h>
+#include <memory>
+
+#if USE_KRB5
+#   include <gssapi/gssapi.h>
+#   include <gssapi/gssapi_ext.h>
+#   define MAYBE_NORETURN
+#else
+#   define MAYBE_NORETURN [[noreturn]]
+#endif
+
+namespace Poco { class Logger; }
+
+namespace DB
+{
+
+class GSSAcceptorContext
+    : public Credentials
```

Comment:
I suppose it will be better without being inherited from `Credentials`. I mean it's GSS acceptor context, it helps up to get credentials, ok but it's not the credentials itself. It seems all the context-related stuff (i.e. the handles and the flag `is_failed`) is used only by `HTTPHandler`, so it could be done easily.

Reply:
It does contain credentials once all tokens from the user are processed. The acceptor context is a "container" and for credentials basically, once we feed all the tokens from user into it and they are decoded.

Reply:
I understand that it's a container. It just seems to me the code would be clearer if you make it a container, not a subclass:

```
class GSSAcceptorContext
{
public:
  // Returns credentials if ready.
  Credentials getCredentials() const;

private:
  Credentials credentials;
  bool ready;
};
```

Reply:
The entire pipeline relies on it being a `Credentials`, because it is stored and updated before being checked. Otherwise, a very custom code will have to be written to store GSSAcceptorContext in HTTPHandler.
Maybe the fact that it is called GSSAcceptorContext and not GSSCredentials causes all this confusion?..

Reply:
I don't see why the pipeline relies on it being a Credentials. It seems they can be separated from each other easily, just by adding two lines to `HTTPHandler`:

```
if (!gss_acceptor_context->isFailed() && !gss_acceptor_context->isReady())
{
  ...
}
+else if (gss_acceptor_context->isReady())
+{
+  request_credentials = gss_acceptor_context->getCredentials();
+}
```

Reply:
Frankly, I don't see any point in doing this, the internal GSS code is tightly coupled, so `credentials` instance would have to stay pretty much just empty, with a single flag. We don't need just another concept floating around the code, where it can be bounded withing the Credential concept.